[flang] Use saturated intrinsic for floating point conversions

ashermancinelli opened the pull request:
The saturated floating point conversion intrinsics more closely match the semantics in the standard.

Case 2 of 16.9.100 is

> INT (A [, KIND])
> If A is of type real, there are two cases: if |A| < 1, INT (A) has the value 0; if |A| ≥ 1, INT (A) is the integer whose magnitude is the largest integer that does not exceed the magnitude of A and whose sign is the same as the sign of A.

Currently, converting a floating point value into an integer type too small to hold the constant will be converted to poison in opt, leaving us with garbage:

```
> cat t.f90
program main
  real(kind=16)   :: f
  integer(kind=4) :: i
  f=huge(f)
  i=f
  print *, i
end program main

# current upstream
> for i in `seq 10`; do; ./a.out; done
 -862156992
 -1497393344
 -739096768
 -1649494208
 1761228608
 -1959270592
 -746244288
 -1629194432
 -231217344
 382322496
```

With the saturated fptoui/fptosi intrinsics, we get the appropriate values

```
# mine
> flang -O2 ./t.f90 && ./a.out
 2147483647

> perl -e 'printf "%d\n", (2 ** 31) - 1'
2147483647
```

One notable difference: NaNs being converted to ints will become zero, unlike current flang (and other compilers).

ashermancinelli committed:
[fircg] Lower to saturated fp to int conversions

fptosi and fptoui llvm instructions become poison during optimization
when the conversion cannot be performed.

The standard mandates that a real converted to an int must be converted
to the largest integer that does not exceed the magnitude of the original
and keeps the same sign. The saturated floating point conversions match
these semantics more closely than the regular conversion instructions.

ashermancinelli committed:
Add unsigned conversion checks

ashermancinelli commented:
It looks like newer versions of gfortran actually have the same NaN -> int behavior that we will have if this patch is merged:
```
> gfortran --version           
GNU Fortran (GCC) 9.3.0
Copyright (C) 2019 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

> gfortran -O3 t.f90 && ./a.out
 -2147483648

> gfortran --version
GNU Fortran (GCC) 14.1.0
Copyright (C) 2024 Free Software Foundation, Inc.
This is free software; see the source for copying conditions.  There is NO
warranty; not even for MERCHANTABILITY or FITNESS FOR A PARTICULAR PURPOSE.

> gfortran -O3 t.f90 && ./a.out
           0
```
```
program main
  use ieee_arithmetic, only: ieee_value, ieee_quiet_nan
  implicit none
  real(kind=16)   :: f
  integer(kind=4) :: i
  f=ieee_value(f, ieee_quiet_nan)
  i=f
  print *, i
end program main
```

jeanPerier commented on flang/test/Integration/fp-convert.f90:
```diff
@@ -0,0 +1,236 @@
+! RUN: %flang -funsigned %s -o %t && %t | FileCheck %s
```
Lit tests should not contain end-to-end tests, they belong to https://github.com/llvm/llvm-test-suite/tree/main/Fortran/UnitTests (this is a "whole program test" for https://llvm.org/docs/TestingGuide.html#llvm-testing-infrastructure-organization).

It is a great end to end test, so I would encourage you to add it there.

ashermancinelli replied:
Thanks! Do you think the small changes I made to `flang/test/Fir/convert-to-llvm.fir` are enough for this PR, then? Those are the only differences we should see at the fir level.

jeanPerier commented on flang/test/Integration/fp-convert.f90:
```diff
@@ -0,0 +1,236 @@
+  end interface
+contains
+  subroutine set_and_print_r16(value)
+    real(kind=16), intent(in) :: value
```
real 16 is not enabled on all targets, even for X86-64, it is not always enabled (hence the buildbot failure), you can workaround this by using merge + target dependent check as it was done in https://github.com/llvm/llvm-project/commit/1128343727ce17f65976ef791783cd737f8cb5bd

jeanPerier reviewed (changes requested):
Thanks Asher, aside from the test issues, looks good to me.
I added a few other reviewers working on other platforms for more visibility.

kiranchandramohan commented on flang/lib/Optimizer/CodeGen/CodeGen.cpp:
```diff
@@ -835,10 +835,20 @@ struct ConvertOpConversion : public fir::FIROpConversion<fir::ConvertOp> {
+        // NOTE: We are checking the fir type here because toTy is an LLVM type
+        // which is signless, and we need to use the intrinsic that matches the
+        // sign of the output in fir.
+        if (toFirTy.isUnsignedInteger()) {
+          auto intrinsicName =
+              mlir::StringAttr::get(convert.getContext(), "llvm.fptoui.sat");
+          rewriter.replaceOpWithNewOp<mlir::LLVM::CallIntrinsicOp>(
+              convert, toTy, intrinsicName, op0);
+        } else {
+          auto intrinsicName =
+              mlir::StringAttr::get(convert.getContext(), "llvm.fptosi.sat");
+          rewriter.replaceOpWithNewOp<mlir::LLVM::CallIntrinsicOp>(
+              convert, toTy, intrinsicName, op0);
+        }
```
Would you know what these saturation intrinsics get lowered to? And is there a big difference in performance? Would it be possible to use the saturation intrinsic only when necessary? Or can that not be determined at compile time?

ashermancinelli replied:
They produce more instructions on x86 (when they cannot be const-folded away) ([x86 godbolt link, more instructions](https://godbolt.org/z/z6KKf8Yao), [aarch64 godbolt link, both using `fcvtzs`](https://godbolt.org/z/7coacsjPd)), and if someone converted reals to integers in a hot loop they might see worse performance, however I was unable to find a difference in the performance tests that I ran. I'll be watching performance numbers after this is merged in case something comes up. 

> Would it be possible to use the saturation intrinsic only when necessary?

As long as we want the correct semantics for values only known at runtime, I don't think so. However, especially if performance issues come up, I think it would make sense to use the fptosi/fptoui instructions under some flag, maybe enabled by default above some optimization level. Do you think using the instructions instead of the saturated intrinsics under (for example) `-ffast-math` would be a good compromise if performance issues show up?

jeanPerier replied:
> Do you think using the instructions instead of the saturated intrinsics under (for example) -ffast-math would be a good compromise if performance issues show up?

Personally, I agree with that approach. I think it is better to avoid having too many code generation paths unless there is an actual use case for it, in which case ` -ffast-math` would sounds like the right flag to deviate from the requirements.

Please wait for @kiranchandramohan's feedback on the matter.

kiranchandramohan replied:
Thanks @ashermancinelli for the reply. Just a few points, thinking out loud.

> Would it be possible to use the saturation intrinsic only when necessary? Or can that not be determined at compile time?

This question I was asking here was about inferring from the real and integer types involved in the conversion. Like if we are converting from real(kind=2)/half-precision to integer(kind=4) then probably integer(kind=4) can hold all values without saturation.

gfortran (without fast-math) seems to be calling `__fixtfsi`. 

There is also a question of whether vectorisation will work for these saturation intrinsics. I can see one issue filed against this topic by the rust community.
https://github.com/llvm/llvm-project/issues/59682

> > Do you think using the instructions instead of the saturated intrinsics under (for example) -ffast-math would be a good compromise if performance issues show up?
> 
> Personally, I agree with that approach. I think it is better to avoid having too many code generation paths unless there is an actual use case for it, in which case ` -ffast-math` would sounds like the right flag to deviate from the requirements.
> 
> Please wait for @kiranchandramohan's feedback on the matter.

Makes sense.

ashermancinelli replied:
> Like if we are converting from real(kind=2)/half-precision to integer(kind=4) then probably integer(kind=4) can hold all values without saturation.

I see what you mean now and that seems like a great idea. I see you've approved this PR so I'll merge for now, but I would like to address this in a follow-up. Thanks!

ashermancinelli committed:
Remove end-to-end test case

As Jean pointed out, this is better-suited to llvm-test-suite, so I'll
add it there after this change is merged.

ashermancinelli commented:
I'll add the end-to-end test to the llvm test suite once this is merged. Thanks for the reviews!

kiranchandramohan reviewed (approved):
LG.
